R/keyword_quote_linter.R

```diff
@@ -0,0 +1,129 @@
+    | //OP-AT/following-sibling::SLOT[starts-with(text(), '`')]
+  "
+
+  return(Linter(function(source_expression) {
+    if (!is_lint_level(source_expression, "expression")) {
+      return(list())
+    }
+
+    xml <- source_expression$xml_parsed_content
+
+    call_arg_expr <- xml_find_all(xml, call_arg_xpath)
+
+    invalid_call_quoting <- is_valid_r_name(get_r_string(call_arg_expr))
+
+    call_arg_lints <- xml_nodes_to_lints(
+      call_arg_expr[invalid_call_quoting],
+      source_expression = source_expression,
+      lint_message = "Only quote named arguments to functions if necessary, i.e., the name is not a valid R symbol (see ?make.names).",
+      type = "warning"
+    )
+
+    assignment_expr <- xml_find_all(xml, assignment_xpath)
+
+    invalid_assignment_quoting <- is_valid_r_name(get_r_string(assignment_expr), no_quote = TRUE)
+
+    assignment_lints <- xml_nodes_to_lints(
+      assignment_expr[invalid_assignment_quoting],
+      source_expression = source_expression,
+      lint_message = paste(
+        "Only quote targets of assignment if necessary, i.e., the name is not a valid R symbol (see ?make.names).",
+        "If necessary, use backticks to create non-syntactic names, not quotes."
+      ),
+      type = "warning"
+    )
+
+    extraction_expr <- xml_find_all(xml, extraction_xpath)
+
+    invalid_extraction_quoting <-
+      is_valid_r_name(get_r_string(extraction_expr), no_quote = TRUE)
+
+    extraction_expr <- extraction_expr[invalid_extraction_quoting]
+    extractor <- xml_find_chr(extraction_expr, "string(preceding-sibling::*[1])")
+    gen_extractor <- ifelse(extractor == "$", "[[", "slot()")
+
+    extraction_lints <- xml_nodes_to_lints(
+      extraction_expr[invalid_extraction_quoting],
+      source_expression = source_expression,
+      lint_message = paste(
+        "Only quote targets of extraction with", extractor, "if necessary, i.e., the name is not a valid R symbol (see ?make.names).",
+        "If necessary, use backticks to create non-syntactic names, not quotes, or use", gen_extractor, "to extract by string."
+      ),
+      type = "warning"
+    )
+
+    return(c(call_arg_lints, assignment_lints, extraction_lints))
+  }))
+}
+
+# from ?Reserved
+r_reserved_words <- c(
+  "if", "else", "repeat", "while", "function", "for", "in", "next", "break",
+  "TRUE", "FALSE", "NULL", "Inf", "NaN", "NA",
+  "NA_integer_", "NA_real_", "NA_complex_", "NA_character_"
+)
+
+#' Check if a string could be assigned as an R variable.
+#'
+#'   considered valid, i.e., anything wrapped in '' or "" is linted.
+#'
+#' See [make.names()] for the description of syntactically valid names in R.
+#'   we could also replace this with `make.names(x) == x`
+#' @noRd
+is_valid_r_name <- function(x, no_quote = FALSE) {
+  if (no_quote) {
+    bad_quote <- !startsWith(x, "`")
+  } else {
+    bad_quote <- FALSE
+  }
+  is_valid_symbol <- grepl("^([a-zA-Z][a-zA-Z0-9._]*|[.]|[.][a-zA-Z._][a-zA-Z0-9._]*)$", x)
```

Comment:
This regex can be shortened to

`^([a-zA-Z]|[.]$|[.][a-zA-Z._])[a-zA-Z0-9._]*$`

Reply:
that nested `$` is messy... let's just use `make.names()` to offshore the logic.